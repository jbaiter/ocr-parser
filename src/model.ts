export type PageLocation = Dimensions & Coordinates & {
  /** Polygon defining the outline of the element. */
  polygon?: Polygon;
}
/** Dimensions of an element's bounding box, expressed in pixels */
export type Dimensions = {
  width: number;
  height: number;
};

/**
 * Coordinates of a point on the page, expressed in pixel units with
 * the origin in the upper left. If set on an element itself, the point
 * is the upper-left corner of the bounding box.
 */
export type Coordinates = {
  x: number;
  y: number;
};

/** Polygon defining the outlines of an element. */
export type Polygon = Coordinates[];

/** A choice among multiple alternative words for a given position. */
export type WordChoice = {
  text: string;

  /** Floating point number between 0 and 1 */
  probability?: number;
};

/**
 * A 'word' as recognized by an OCR engine.
 *
 * Corresponds to the following types:
 *
 * - HOCR: `ocrx_word`
 * - ALTO: `String` and descendants
 */
export type OcrWord = {
    type: 'word';
    text: string;
    /** Floating point number between 0 and 1 */
    confidence?: number;
    /** Multiple alternative readings for this word. */
    choices?: WordChoice[];
    /** Is the word the first part of a hyphenated word? */
    hyphenStart?: boolean;
  } & PageLocation;

/**
 * A line as recognized by an OCR engine.
 *
 * Corresponds to the following types:
 *
 * - HOCR: `ocr_line`
 * - ALTO: `TextLine`
 */
export type OcrLine = {
    type: 'line';
    /**
     * Baseline upon which the characters in the word rest. Defined by a set of
     * coordinates that form a polyline.
     */
    baseline?: Coordinates[];
    /** Words on the line */
    words: OcrWord[];
    /** Words intermingled with text outside of words (including spaces) */
    children: (OcrWord | string)[];
    /** Plaintext of the line */
    text: string;
  } & PageLocation;

export function makeLine(
  spec: Omit<OcrLine, 'type' | 'children' | 'words' | 'text'>,
): OcrLine {
  return {
    ...spec,
    type: 'line',
    children: [],
    get words(): OcrWord[] {
      return this.children.filter((c): c is OcrWord => typeof c !== 'string');
    },
    get text(): string {
      return this.children
        .map((c) => (typeof c === 'string' ? c : c.text))
        .join('');
    },
  };
}

/**
 * A paragraph as recognized by an OCR engine.
 *
 * Corresponds to the following types:
 *
 * - HOCR: `ocr_par`
 * - ALTO: Not mapped
 */
export type OcrParagraph = {
  type: 'paragraph';
  /** Descendant elements of the paragraph, currently only lines possible. */
  children: OcrLine[];
  /** All lines in the paragraph. */
  lines: OcrLine[];
  /** All words in the paragraph. */
  words: OcrWord[];
  /** Text content of the paragraph. */
  text: string;
} & Partial<PageLocation>;

export function makeParagraph(
  spec: Omit<OcrParagraph, 'type' | 'children' | 'lines' | 'words' | 'text'>,
): OcrParagraph {
  return {
    ...spec,
    type: 'paragraph',
    children: [],
    get lines(): OcrLine[] {
      return this.children;
    },
    get words(): OcrWord[] {
      return this.children.flatMap((l) => l.words);
    },
    get text(): string {
      const out: string[] = [];
      for (const [idx, child] of this.children.entries()) {
        out.push(child.text);
        if (
          idx < this.children.length - 1 &&
          !child.words.slice(-1)[0]?.hyphenStart
        ) {
          out.push(' ');
        }
      }
      return out.join('');
    },
  };
}

/**
 * A 'block' of text as recognized by an OCR engine.
 *
 * Corresponds to the following types:
 *
 * - HOCR: `ocr_carea`
 * - ALTO: `TextBlock`
 */
export type OcrBlock = {
    type: 'block';
    /** Descendant elements of the block, can be paragraphs or lines. */
    children: (OcrParagraph | OcrLine)[];
    /** Paragraphs in the block */
    paragraphs: OcrParagraph[];
    /** All lines in the block. */
    lines: OcrLine[];
    /** All words in the block. */
    words: OcrWord[];
    /** Text content of the block. */
    text: string;
  } & PageLocation;

export function makeBlock(
  spec: Omit<
    OcrBlock,
    'type' | 'children' | 'paragraphs' | 'lines' | 'words' | 'text'
  >,
): OcrBlock {
  return {
    ...spec,
    type: 'block',
    children: [],
    get paragraphs(): OcrParagraph[] {
      return this.children.filter((c): c is OcrParagraph => 'lines' in c);
    },
    get lines(): OcrLine[] {
      return this.children.reduce((acc, c) => {
        if ('lines' in c) {
          acc.push(...c.lines);
        } else if ('words' in c) {
          acc.push(c);
        }
        return acc;
      }, [] as OcrLine[]);
    },
    get words(): OcrWord[] {
      return this.lines.flatMap((l) => l.words);
    },
    get text(): string {
      const out: string[] = [];
      for (const [idx, child] of this.children.entries()) {
        out.push(child.text);
        if (
          idx < this.children.length - 1 &&
          !child.words.slice(-1)[0]?.hyphenStart
        ) {
          if (child.type === 'line') {
            out.push(' ');
          } else {
            out.push('\n');
          }
        }
      }
      return out.join('');
    },
  };
}

const OCR_FEATURES = {
  POLYGONS: 'POLYGONS',
  BASELINE: 'BASELINE',
  CONFIDENCE: 'CONFIDENCE',
  CHOICES: 'CHOICES',
  HYPHEN: 'HYPHEN',
} as const;
type OcrFeature = typeof OCR_FEATURES[keyof typeof OCR_FEATURES];

type ImageSource = {
  /** File name or path associated with the image the page was sourced from. */
  fileName?: string;
  /** Generic identifier for the image file the page was sourced from. */
  fileIdentifier?: string;
  /** Generic identifier for the document the page was souced from. */
  documentIdentifier?: string;
  /** Resolution in pixels per inch of the scanned document image the page was sourced from. */
  ppi?: number;
  /** Checksum of the associated image file. */
  checksum?: string;
  /** Type of algorithm used to calculate the checksum. */
  checksumType?: string;
}

/**
 * A page of text as recognized by an OCR engine.
 *
 * Corresponds to the following types:
 *
 * - HOCR: `ocr_page`
 * - ALTO: `Page`
 */
export type OcrPage = Dimensions & {
  type: 'page';
  /** Special features used by this page. */
  features: OcrFeature[];
  /** Information about associated image. */
  imageSource?: ImageSource;
  /** Physical page number of the page. */
  physicalPageNumber?: number;
  /** "Logical" page number as printed/written on the page */
  logicalPageNumber?: string;
  /** Direct descendants of the page element. */
  children: (OcrBlock | OcrParagraph | OcrLine)[];
  /** All blocks of text on the page */
  blocks: OcrBlock[];
  /** All paragraphs on the page. */
  paragraphs: OcrParagraph[];
  /** All lines on the page. */
  lines: OcrLine[];
  /** All words on the page. */
  words: OcrWord[];
  /** Text content of the page. */
  text: string;
};

export function makePage(
  spec: Omit<
    OcrPage,
    'type' | 'children' | 'blocks' | 'paragraphs' | 'lines' | 'words' | 'text'
  >,
): OcrPage {
  return {
    ...spec,
    type: 'page',
    children: [],
    get blocks(): OcrBlock[] {
      return this.children.filter((c): c is OcrBlock => 'paragraphs' in c);
    },
    get paragraphs(): OcrParagraph[] {
      return this.children.reduce((acc, c) => {
        if ('paragraphs' in c) {
          acc.push(...c.paragraphs);
        } else if ('lines' in c) {
          acc.push(c);
        }
        return acc;
      }, [] as OcrParagraph[]);
    },
    get lines(): OcrLine[] {
      return this.children.reduce((acc, c) => {
        if ('lines' in c) {
          acc.push(...c.lines);
        } else if ('words' in c) {
          acc.push(c);
        }
        return acc;
      }, [] as OcrLine[]);
    },
    get words(): OcrWord[] {
      return this.children.flatMap((l) => l.words);
    },
    get text(): string {
      const out: string[] = [];
      for (const [idx, child] of this.children.entries()) {
        out.push(child.text);
        if (
          idx < this.children.length - 1 &&
          !child.words.slice(-1)[0]?.hyphenStart
        ) {
          if (child.type === 'line') {
            out.push(' ');
          } else {
            out.push('\n');
          }
        }
      }
      return out.join('');
    },
  };
}

/** Union of all possible OCR element types. */
export type OcrElement =
  | string
  | OcrWord
  | OcrLine
  | OcrParagraph
  | OcrBlock
  | OcrPage;
