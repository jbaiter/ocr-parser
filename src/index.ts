import { parseHocrPages } from './hocr.js';
import { parseAltoPages } from './alto.js';
import { toReadableStream, isInitialized } from './util.js';
import { type Dimensions, type OcrPage } from './model.js';

export { parseHocrPages } from './hocr.js';
export { parseAltoPages } from './alto.js';
export { initialize, isInitialized, setupLogging, SAX_WASM_VERSION } from './util.js';
export type {
  Dimensions,
  Coordinates,
  PageLocation,
  WordChoice,
  OcrPage,
  OcrBlock,
  OcrParagraph,
  OcrLine,
  OcrWord,
  OcrFeature,
  OcrFeatures,
  ImageSource,
  Polygon,
} from './model';

/**
 * Callback to provide a reference size for a given page.
 *
 * @param idx: 0-based index of the page in the document.
 * @param pageAttribs: XML attributes of the page element.
 * @returns The reference size , or `null` if no reference size is available.
 */
export type ReferenceSizeCallback = (
  idx: number,
  pageAttribs: { [key: string]: string },
) => Dimensions | null;

/**
 * Parse a single OCR page in either ALTO or hOCR format.
 *
 * Must have run {@link initialize} first.
 *
 * @param markup The document to parse. Can be a `string`, a `Uint8Array` or a
 *   `ReadableStream`.
 * @param format The format of the document.
 * @param referenceSizes Reference page dimensions These dimensions can be used
 *   to scale all coordinates to a given reference frame in cases when the
 *   display resolution is different from the OCR resolution. In the case of
 *   ALTO, a reference size might be required to obtain pixel coordinates,
 *   since in some documents the coordinates are expressed in non-pixel units.
 * @returns An async generator that yields `OcrPage` objects.
 */
export async function parseOcrPage(
  /** HOCR/ALTO markup for a single page. */
  markup: string | ReadableStream | Uint8Array,
  format: 'hocr' | 'alto',
  referenceSize?: Dimensions,
): Promise<OcrPage> {
  if (!isInitialized()) {
    throw new Error(
      'XML parser WASM not initialized, call initialize() first!',
    );
  }
  let page: OcrPage | null;
  if (format === 'hocr') {
    page = (
      await parseHocrPages(
        markup,
        referenceSize ? [referenceSize] : undefined,
      ).next()
    ).value;
  } else if (format === 'alto') {
    page = (
      await parseAltoPages(
        markup,
        referenceSize ? [referenceSize] : undefined,
      ).next()
    ).value;
  } else {
    throw new Error(`Unsupported format: ${format}`);
  }
  if (!page) {
    throw new Error('Failed to parse page');
  }
  return page;
}

/**
 * Parse all OCR pages in the given hOCR or ALTO markup.
 *
 * Must have run {@link initialize} first.
 *
 * @param markup The document to parse. Can be a `string`, a `Uint8Array` or a
 *   `ReadableStream`.
 * @param format The format of the document.
 * @param referenceSizes An array of dimensions for each page in the document,
 *   or a callback that returns the dimensions for a given page, based on its
 *   numerical index and its XML attributes.These dimensions can be used to
 *   scale all coordinates to a given reference frame in cases when the display
 *   resolution is different from the OCR resolution. In the case of ALTO, a
 *   reference size might be required to obtain pixel coordinates, since in
 *   some documents the coordinates are expressed in non-pixel units.
 * @returns An async generator that yields `OcrPage` objects.
 */
export async function* parseOcrPages(
  markup: string | ReadableStream | Uint8Array,
  format: 'hocr' | 'alto',
  referenceSizes?: Dimensions[] | ReferenceSizeCallback,
): AsyncGenerator<OcrPage> {
  if (!isInitialized()) {
    throw new Error(
      'XML parser WASM not initialized, call initialize() first!',
    );
  }
  if (!(markup instanceof ReadableStream)) {
    markup = toReadableStream(markup);
  }
  switch (format) {
    case 'hocr':
      yield* parseHocrPages(markup, referenceSizes);
      break;
    case 'alto':
      yield* parseAltoPages(markup, referenceSizes);
      break;
  }
}
