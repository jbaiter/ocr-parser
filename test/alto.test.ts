//@ts-expect-error
import WITH_ALTERNATIVES from './__fixtures__/chronicling_america.xml?raw';

import fs from 'fs';
import { describe, it, expect, beforeAll } from 'vitest';
import { parseAltoPages, initialize } from '../src';

const WITH_SHAPE = `
<?xml version="1.0" encoding="UTF-8"?>
<alto xmlns="http://schema.ccs-gmbh.com/ALTO">
  <Description>
    <MeasurementUnit>pixel</MeasurementUnit>
    <sourceImageInformation>
      <fileName>shape.png</fileName>
    </sourceImageInformation>
  </Description>
  <Layout>
    <Page ID="ID1" HEIGHT="2560" WIDTH="1440" PHYSICAL_IMG_NR="1" PRINTED_IMG_NR="I">
      <PrintSpace HEIGHT="24453.0" WIDTH="19500.0" HPOS="294.0" VPOS="954.0" PC="0.93230003">
        <TextBlock ID="TextBlock4" HEIGHT="198" WIDTH="785" HPOS="318" VPOS="1456" >
          <TextLine ID="TextLine8">
            <Shape>
              <Polygon POINTS="325 1456 1023 1456 1023 1504 419 1504 419 1552 325 1552"/>
            </Shape>
            <String ID="String33" CONTENT="EEn" WC="0.82" CC="050">
              <Shape><Polygon POINTS="325 1456 419 1456 419 1552 325 1552"/></shape>
            </String>
            <SP ID="SP26" WIDTH="28" HPOS="472" VPOS="1456"/>
            <String ID="String34" CONTENT="F~iJJcher" HEIGHT="43" WIDTH="152"  HPOS="500" VPOS="1457" WC="0.82" CC="860001000"/>
          </TextLine>
        </TextBlock>
      </PrintSpace>
    </Page>
  </Layout>
</alto>
`;

const saxWasm = fs.readFileSync(
  __dirname + '/../node_modules/sax-wasm/lib/sax-wasm.wasm',
);

// Custom serializer that ignores getters
expect.addSnapshotSerializer({
  test: (val) =>
    typeof val === 'object' &&
    Object.values(Object.getOwnPropertyDescriptors(val)).some((d) => d.get),
  serialize(val, config, indentation, depth, refs, printer) {
    const filtered = Object.fromEntries(
      Object.entries(val).filter(
        ([k]) => !Object.getOwnPropertyDescriptor(val, k)?.get,
      ),
    );
    return printer(filtered, config, indentation, depth, refs);
  },
});

async function toArray<T>(gen: AsyncGenerator<T>): Promise<T[]> {
  const result: T[] = [];
  for await (let page of gen) {
    result.push(page);
  }
  return result;
}

describe('parseAltoPage', () => {
  beforeAll(() => {
    initialize(() => Promise.resolve(new Uint8Array(saxWasm)));
  });
  it('parses alto lines', async () => {
    const pages = await toArray(
      parseAltoPages(WITH_ALTERNATIVES, [
        {
          width: 6113,
          height: 5115,
        },
      ]),
    );
    expect(pages).toHaveLength(1);
    const page = pages[0];
    expect(page?.text).toMatchInlineSnapshot(
      '"J a Ira mj iI tE1r 3 i c JiLas Edition THE WINCHESTER NEWS I her injuries wsis made Dr Reynolds an eye specialist after lite examina cents and GO of Henry C Hall average weight 1410 > ounds at 0 cents"',
    );
    expect(page).toMatchSnapshot();
  });

  it('parses alto lines with shapes', async () => {
    const pages = await toArray(parseAltoPages(WITH_SHAPE));
    expect(pages).toHaveLength(1);
    const page = pages[0];
    expect(page?.text).toEqual('EEn F~iJJcher');
    expect(page).toMatchSnapshot();
  });

  it('parses alto without explicitly encoded spaces', async () => {
    const pages = await toArray(
      parseAltoPages(WITH_ALTERNATIVES.replace(/<SP.*?\/>/g, ''), [
        {
          width: 6113,
          height: 5115,
        },
      ]),
    );
    expect(pages).toHaveLength(1);
    const page = pages[0];
    expect(page?.text).toEqual(
      'J a Ira mj iI tE1r 3 i c JiLas Edition THE WINCHESTER NEWS I her injuries wsis made Dr Reynolds an eye specialist after lite examina cents and GO of Henry C Hall average weight 1410 > ounds at 0 cents',
    );
    expect(page).toMatchSnapshot();
  });
});
