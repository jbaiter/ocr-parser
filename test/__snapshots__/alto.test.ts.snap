// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseAltoPage > parses alto lines 2`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "confidence": 0.8095238,
              "height": 19.08766046027524,
              "text": "J",
              "type": "word",
              "width": 8.483404649011218,
              "x": 933.174511391234,
              "y": 231.17277668555568,
            },
            " ",
            {
              "confidence": 0.8095238,
              "height": 15.552908523187233,
              "text": "a",
              "type": "word",
              "width": 9.89730542384642,
              "x": 950.1413206892564,
              "y": 234.0005782352261,
            },
            " ",
            {
              "confidence": 0.95238096,
              "height": 19.08766046027524,
              "text": "Ira",
              "type": "word",
              "width": 53.72822944373771,
              "x": 976.2984850237077,
              "y": 229.05192552330288,
            },
            " ",
            {
              "confidence": 0.8095238,
              "height": 19.79461084769284,
              "text": "mj",
              "type": "word",
              "width": 19.08766046027524,
              "x": 1135.362322192668,
              "y": 229.7588759107205,
            },
            " ",
            {
              "confidence": 0.8095238,
              "height": 14.845958135769632,
              "text": "iI",
              "type": "word",
              "width": 24.743263559616054,
              "x": 1316.341621371574,
              "y": 231.87972707297328,
            },
            " ",
            {
              "confidence": 0.8095238,
              "height": 21.208511622528047,
              "text": "tE1r",
              "type": "word",
              "width": 190.8766046027524,
              "x": 1439.3509887822368,
              "y": 229.05192552330288,
            },
            " ",
            {
              "confidence": 0.8095238,
              "height": 12.018156586099225,
              "text": "3",
              "type": "word",
              "width": 6.362553486758413,
              "x": 1768.0829189314213,
              "y": 237.5353301723141,
            },
          ],
          "height": 21.208511622528047,
          "type": "line",
          "width": 841.2709610269458,
          "x": 933.174511391234,
          "y": 229.05192552330288,
        },
        {
          "children": [
            {
              "confidence": 0.8095238,
              "height": 9.19035503642882,
              "text": "i",
              "type": "word",
              "width": 5.655603099340812,
              "x": 1600.5356771134498,
              "y": 250.96738753324854,
            },
          ],
          "height": 9.19035503642882,
          "type": "line",
          "width": 5.655603099340812,
          "x": 1600.5356771134498,
          "y": 250.96738753324854,
        },
        {
          "children": [
            {
              "confidence": 0.8095238,
              "height": 9.89730542384642,
              "text": "c",
              "type": "word",
              "width": 7.069503874176015,
              "x": 593.1313750433676,
              "y": 265.10639528160056,
            },
          ],
          "height": 9.89730542384642,
          "type": "line",
          "width": 7.069503874176015,
          "x": 593.1313750433676,
          "y": 265.10639528160056,
        },
        {
          "children": [
            {
              "choices": [
                {
                  "text": "Jails",
                },
              ],
              "confidence": 0.8095238,
              "height": 299.04001387764544,
              "text": "JiLas",
              "type": "word",
              "width": 483.55406499363943,
              "x": 69.28113796692494,
              "y": 330.8527813114375,
            },
            " ",
            {
              "confidence": 0.95238096,
              "height": 51.60737828148491,
              "text": "Edition",
              "type": "word",
              "width": 241.0700821094021,
              "x": 617.1676882155662,
              "y": 514.6598820400139,
            },
            " ",
            {
              "choices": [
                {
                  "text": "TIIE",
                },
                {
                  "text": "TIE",
                },
              ],
              "confidence": 1,
              "height": 181.6862495663236,
              "text": "THE",
              "type": "word",
              "width": 452.448247947265,
              "x": 1018.0085578813462,
              "y": 433.36058748698974,
            },
            " ",
            {
              "confidence": 1,
              "height": 186.63490227824678,
              "text": "WINCHESTER",
              "type": "word",
              "width": 1487.4236151266336,
              "x": 1589.224470914768,
              "y": 431.94668671215453,
            },
            " ",
            {
              "confidence": 1,
              "height": 181.6862495663236,
              "text": "NEWS",
              "type": "word",
              "width": 644.7387533248526,
              "x": 3201.0713542268995,
              "y": 429.8258355499017,
            },
          ],
          "height": 299.04001387764544,
          "type": "line",
          "width": 3776.5289695848273,
          "x": 69.28113796692494,
          "y": 330.8527813114375,
        },
        {
          "children": [
            {
              "confidence": 0.8095238,
              "height": 57.96993176824332,
              "text": "I",
              "type": "word",
              "width": 3.5347519370880076,
              "x": 947.313519139586,
              "y": 766.33421996068,
            },
            " ",
            {
              "confidence": 1,
              "height": 25.450213947033653,
              "text": "her",
              "type": "word",
              "width": 47.3656759569793,
              "x": 962.1594772753556,
              "y": 770.5759222851856,
            },
            " ",
            {
              "choices": [
                {
                  "text": "inlurie",
                },
                {
                  "text": "inure",
                },
              ],
              "confidence": 0.95238096,
              "height": 31.105817046374465,
              "text": "injuries",
              "type": "word",
              "width": 113.11206198681624,
              "x": 1022.2502602058518,
              "y": 770.5759222851856,
            },
            " ",
            {
              "choices": [
                {
                  "text": "W1S",
                },
                {
                  "text": "ams",
                },
                {
                  "text": "sis",
                },
                {
                  "text": "aims",
                },
              ],
              "confidence": 0.82539684,
              "height": 16.966809298022437,
              "text": "wsis",
              "type": "word",
              "width": 57.96993176824332,
              "x": 1148.0874291661848,
              "y": 779.7662773216144,
            },
            " ",
            {
              "choices": [
                {
                  "text": "mncle",
                },
                {
                  "text": "manacle",
                },
              ],
              "confidence": 1,
              "height": 25.450213947033653,
              "text": "made",
              "type": "word",
              "width": 77.05759222851856,
              "x": 1220.9033190701978,
              "y": 772.6967734474384,
            },
            " ",
            {
              "choices": [
                {
                  "text": "Ik",
                },
                {
                  "text": "Ir",
                },
              ],
              "confidence": 0.82539684,
              "height": 24.743263559616054,
              "text": "Dr",
              "type": "word",
              "width": 39.58922169538568,
              "x": 1337.5501329941021,
              "y": 772.6967734474384,
            },
            " ",
            {
              "choices": [
                {
                  "text": "Reynold",
                },
              ],
              "confidence": 0.95238096,
              "height": 31.105817046374465,
              "text": "Reynolds",
              "type": "word",
              "width": 134.32057360934428,
              "x": 1399.0548166994333,
              "y": 772.6967734474384,
            },
          ],
          "height": 57.96993176824332,
          "type": "line",
          "width": 586.0618711691916,
          "x": 947.313519139586,
          "y": 766.33421996068,
        },
        {
          "children": [
            {
              "confidence": 0.82539684,
              "height": 16.259858910604834,
              "text": "an",
              "type": "word",
              "width": 33.226668208627274,
              "x": 960.7455765005204,
              "y": 817.9415982421649,
            },
            " ",
            {
              "confidence": 1,
              "height": 22.62241239736325,
              "text": "eye",
              "type": "word",
              "width": 48.7795767318145,
              "x": 1013.0599051694229,
              "y": 817.9415982421649,
            },
            " ",
            {
              "choices": [
                {
                  "text": "pliitlist",
                },
                {
                  "text": "littlest",
                },
              ],
              "confidence": 0.95238096,
              "height": 31.105817046374465,
              "text": "specialist",
              "type": "word",
              "width": 140.6831270961027,
              "x": 1078.80629119926,
              "y": 810.1651439805713,
            },
            " ",
            {
              "choices": [
                {
                  "text": "aftei",
                },
              ],
              "confidence": 0.95238096,
              "height": 24.03631317219845,
              "text": "after",
              "type": "word",
              "width": 75.64369145368336,
              "x": 1243.525731467561,
              "y": 812.9929455302417,
            },
            " ",
            {
              "choices": [
                {
                  "text": "tlle",
                },
                {
                  "text": "elite",
                },
                {
                  "text": "tulle",
                },
              ],
              "confidence": 0.82539684,
              "height": 28.98496588412166,
              "text": "lite",
              "type": "word",
              "width": 44.53787440730889,
              "x": 1336.1362322192667,
              "y": 808.0442928183185,
            },
            " ",
            {
              "choices": [
                {
                  "text": "exalnina",
                },
                {
                  "text": "examine",
                },
                {
                  "text": "exanimate",
                },
              ],
              "confidence": 0.82539684,
              "height": 25.450213947033653,
              "text": "examina",
              "type": "word",
              "width": 120.18156586099225,
              "x": 1397.6409159245982,
              "y": 812.2859951428242,
            },
          ],
          "height": 33.226668208627274,
          "type": "line",
          "width": 557.07690528507,
          "x": 960.7455765005204,
          "y": 808.0442928183185,
        },
        {
          "children": [
            {
              "confidence": 1,
              "height": 21.208511622528047,
              "text": "cents",
              "type": "word",
              "width": 76.35064184110097,
              "x": 955.7969237885973,
              "y": 1266.1481438649244,
            },
            " ",
            {
              "choices": [
                {
                  "text": "ancl",
                },
                {
                  "text": "8111",
                },
                {
                  "text": "encl",
                },
              ],
              "confidence": 0.95238096,
              "height": 24.743263559616054,
              "text": "and",
              "type": "word",
              "width": 52.31432866890251,
              "x": 1055.476928414479,
              "y": 1263.3203423152538,
            },
            " ",
            {
              "choices": [
                {
                  "text": "GII",
                },
              ],
              "confidence": 0.96825397,
              "height": 24.743263559616054,
              "text": "GO",
              "type": "word",
              "width": 31.812767433792068,
              "x": 1123.3441656065688,
              "y": 1263.3203423152538,
            },
            " ",
            {
              "confidence": 1,
              "height": 25.450213947033653,
              "text": "of",
              "type": "word",
              "width": 31.105817046374465,
              "x": 1170.709841563548,
              "y": 1262.6133919278363,
            },
            " ",
            {
              "choices": [
                {
                  "text": "Hcl1IY",
                },
                {
                  "text": "Ilenly",
                },
                {
                  "text": "Silently",
                },
              ],
              "confidence": 0.95238096,
              "height": 30.398866658956866,
              "text": "Henry",
              "type": "word",
              "width": 90.489649589453,
              "x": 1216.661616745692,
              "y": 1262.6133919278363,
            },
            " ",
            {
              "confidence": 0.96825397,
              "height": 24.743263559616054,
              "text": "C",
              "type": "word",
              "width": 19.08766046027524,
              "x": 1321.2902740834973,
              "y": 1261.9064415404187,
            },
            " ",
            {
              "choices": [
                {
                  "text": "lIall",
                },
                {
                  "text": "hail",
                },
                {
                  "text": "leally",
                },
              ],
              "confidence": 0.95238096,
              "height": 24.03631317219845,
              "text": "Hall",
              "type": "word",
              "width": 69.28113796692494,
              "x": 1365.1211981033885,
              "y": 1262.6133919278363,
            },
            " ",
            {
              "choices": [
                {
                  "text": "1el",
                },
              ],
              "confidence": 1,
              "height": 16.966809298022437,
              "hyphenStart": true,
              "text": "aver",
              "type": "word",
              "width": 65.03943564241933,
              "x": 1449.248294206083,
              "y": 1269.6828958020124,
            },
          ],
          "height": 31.105817046374465,
          "type": "line",
          "width": 558.4908060599051,
          "x": 955.7969237885973,
          "y": 1261.9064415404187,
        },
        {
          "children": [
            {
              "choices": [
                {
                  "text": "agc",
                },
                {
                  "text": "ac",
                },
              ],
              "confidence": 1,
              "height": 21.208511622528047,
              "text": "age",
              "type": "word",
              "width": 50.19347750664971,
              "x": 955.0899734011796,
              "y": 1309.2721174973979,
            },
            " ",
            {
              "confidence": 1,
              "height": 28.98496588412166,
              "text": "weight",
              "type": "word",
              "width": 94.024401526541,
              "x": 1024.3711113681045,
              "y": 1301.4956632358044,
            },
            " ",
            {
              "choices": [
                {
                  "text": "14111",
                },
              ],
              "confidence": 0.82539684,
              "height": 22.62241239736325,
              "text": "1410",
              "type": "word",
              "width": 75.64369145368336,
              "x": 1140.3109749045911,
              "y": 1302.2026136232219,
            },
            " ",
            {
              "confidence": 0.82539684,
              "height": 14.845958135769632,
              "text": ">",
              "type": "word",
              "width": 6.362553486758413,
              "x": 1245.6465826298138,
              "y": 1309.2721174973979,
            },
            " ",
            {
              "choices": [
                {
                  "text": "bounds",
                },
              ],
              "confidence": 0.82539684,
              "height": 24.743263559616054,
              "text": "ounds",
              "type": "word",
              "width": 86.95489765236498,
              "x": 1254.8369376662426,
              "y": 1300.7887128483867,
            },
            " ",
            {
              "confidence": 1,
              "height": 22.62241239736325,
              "text": "at",
              "type": "word",
              "width": 31.105817046374465,
              "x": 1359.4655950040476,
              "y": 1302.9095640106395,
            },
            " ",
            {
              "choices": [
                {
                  "text": "6",
                },
                {
                  "text": "G",
                },
              ],
              "confidence": 0.82539684,
              "height": 24.03631317219845,
              "text": "0",
              "type": "word",
              "width": 14.845958135769632,
              "x": 1408.2451717358622,
              "y": 1300.7887128483867,
            },
            " ",
            {
              "choices": [
                {
                  "text": "ccnts",
                },
                {
                  "text": "accents",
                },
              ],
              "confidence": 1,
              "height": 23.329362784780848,
              "text": "cents",
              "type": "word",
              "width": 74.93674106626575,
              "x": 1442.8857407193248,
              "y": 1301.4956632358044,
            },
          ],
          "height": 29.691916271539263,
          "type": "line",
          "width": 562.7325083844108,
          "x": 955.0899734011796,
          "y": 1300.7887128483867,
        },
      ],
      "height": 400.84086966578,
      "type": "block",
      "width": 3776.5289695848273,
      "x": 69.28113796692494,
      "y": 229.05192552330288,
    },
  ],
  "features": [
    "CHOICES",
    "HYPHEN",
    "CONFIDENCE",
  ],
  "height": 6113,
  "imageSource": {
    "fileName": "
        \\\\\\\\libdpsrv1\\\\storage\\\\root\\\\projects\\\\NDNP\\\\win_19101017-19101231\\\\win_19101017-19101231-scan-1-\\\\ocr\\\\0070.tif",
  },
  "logicalPageNumber": "[1]",
  "physicalPageNumber": 1,
  "type": "page",
  "width": 4822.10859257546,
}
`;

exports[`parseAltoPage > parses alto lines with shapes 2`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "confidence": 0.82,
              "height": 96,
              "polygon": [
                {
                  "x": 325,
                  "y": 1456,
                },
                {
                  "x": 419,
                  "y": 1456,
                },
                {
                  "x": 419,
                  "y": 1552,
                },
                {
                  "x": 325,
                  "y": 1552,
                },
              ],
              "text": "EEn",
              "type": "word",
              "width": 94,
              "x": 325,
              "y": 1456,
            },
          ],
          "height": 96,
          "polygon": [
            {
              "x": 325,
              "y": 1456,
            },
            {
              "x": 1023,
              "y": 1456,
            },
            {
              "x": 1023,
              "y": 1504,
            },
            {
              "x": 419,
              "y": 1504,
            },
            {
              "x": 419,
              "y": 1552,
            },
            {
              "x": 325,
              "y": 1552,
            },
          ],
          "type": "line",
          "width": 698,
          "x": 325,
          "y": 1456,
        },
      ],
      "height": 198,
      "type": "block",
      "width": 785,
      "x": 318,
      "y": 1456,
    },
  ],
  "features": [
    "POLYGONS",
    "CONFIDENCE",
  ],
  "height": 2560,
  "imageSource": {
    "fileName": "shape.png",
  },
  "logicalPageNumber": "I",
  "physicalPageNumber": 1,
  "type": "page",
  "width": 1440,
}
`;
